src/main/java/com/oath/halodb/CompactionManager.java

```diff
@@ -306,18 +310,33 @@ private boolean isRecordFresh(IndexFileEntry entry, InMemoryIndexMetaData metaDa
         }
 
         private void rollOverCurrentWriteFile(int recordSize) throws IOException {
-            if (currentWriteFile == null ||  currentWriteFileOffset + recordSize > dbInternal.options.getMaxFileSize()) {
-                if (currentWriteFile != null) {
-                    currentWriteFile.flushToDisk();
-                    currentWriteFile.getIndexFile().flushToDisk();
+            // Make sure there's no unnecessary roll over of the write
+            compactionFileRollOverLock.lock();
```

Comment:
This is a critical path which each write will go through. This lock may result unnecessary cost. Performance test is needed. Can we find another way to do it more efficiently?

Reply:
Looks like you have used lock in ```forceRolloverCurrentWriteFile```. Is this lock still necessary?

Reply:
There are three files are supposed to be actively writing when roll over happened. So each file Rollover need a separate Lock for that.

Those Reenentry Locks should be very light weight as it only get blocked when two threads are accessing the Locks. From my test in bucket host, there's no system stunk when rollover happened.